core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -65,9 +68,167 @@ public CachedConnection reserveAny() {
       }
       return cachedConnection;
     }
+
+    private void removeExpiredConnections(final ArrayList<CachedConnection> expired, final long killTime) {
+      long currTime = System.currentTimeMillis();
+      while (isLastUnreservedExpired(currTime, killTime)) {
+        expired.add(unreserved.removeLast());
+      }
+    }
+
+    boolean isLastUnreservedExpired(final long currTime, final long killTime) {
+      return !unreserved.isEmpty() && (currTime - unreserved.peekLast().lastReturnTime) > killTime;
+    }
+
+    void checkReservedForStuckIO() {
+      reserved.values().forEach(c -> c.transport.checkForStuckIO(STUCK_THRESHOLD));
+    }
+
+    void closeAllTransports() {
+      closeTransports(unreserved);
+      closeTransports(reserved.values());
+    }
+
+    void closeTransports(final Iterable<CachedConnection> stream) {
+      stream.forEach((connection) -> {
+        try {
+          connection.transport.close();
+        } catch (Exception e) {
+          log.debug("Error closing transport during shutdown", e);
+        }
+      });
+    }
+
+    CachedConnection removeReserved(CachedTTransport transport) {
+      return reserved.remove(transport);
+    }
+  }
+
+  private static class ConnectionPool {
+    final ConcurrentHashMap<ThriftTransportKey,ReentrantLock> locks = new ConcurrentHashMap<>();
+    final ConcurrentHashMap<ThriftTransportKey,CachedConnections> connections = new ConcurrentHashMap<>();
+
+    Set<ThriftTransportKey> getThriftTransportKeys() {
+      return connections.keySet();
+    }
+
+    CachedConnection reserveAny(final ThriftTransportKey key) {
+      return executeWithinLock(key, () -> getOrCreateCachedConnections(key).reserveAny());
+    }
+
+    CachedConnection reserveAnyIfPresent(final ThriftTransportKey key) {
+      return executeWithinLock(key, () -> connections.containsKey(key)
+          ? connections.get(key).reserveAny()
+          : null);
+    }
+
+    void putReserved(final ThriftTransportKey key, final CachedConnection connection) {
+      executeWithinLock(key, () -> getOrCreateCachedConnections(key).reserved.put(
+          connection.transport, connection));
+    }
+
+    boolean returnTransport(final CachedTTransport transport, final List<CachedConnection> toBeClosed) {
+      return executeWithinLock(transport.getCacheKey(), () -> unreserveConnection(transport, toBeClosed));
+    }
+
+    void shutdown() {
+      for (Entry<ThriftTransportKey,CachedConnections> entry : connections.entrySet()) {
```

Comment:
I think there are race conditions where this loop will not see all of the locks (like locks added during the lifetime of the loop).  That is why I used an array of locks in lbschanno#3

Reply:
Agreed, I merged in your changes with the array of locks.